### PR TITLE
Support broadcasting 0-length dimensions

### PR DESCRIPTION
**Pull request by jakirkham**
Fixes https://github.com/dask/dask/issues/2780

Allows 0-length dimensions to be broadcasted and forces the broadcasted shape to 0 along those dimensions. Provides some tests for broadcasting arrays with 0-length dimensions.

### CODE REVIEW

**Comment by jakirkham** [comment]
Not sure if we need to be doing something with chunking too, but this seems to work ok thus far.

**Comment by jakirkham** [comment]
Any thoughts on this?

File: dask/array/core.py [hunk 1]
```diff
@@ -2497,8 +2497,8 @@ def broadcast_shapes(*shapes):
         return shapes[0]
     out = []
     for sizes in zip_longest(*map(reversed, shapes), fillvalue=-1):
-        dim = max(sizes)
-        if any(i != -1 and i != 1 and i != dim and not np.isnan(i) for i in sizes):
+        dim = 0 if 0 in sizes else max(sizes)
+        if any(i != -1 and i != 0 and i != 1 and i != dim and not np.isnan(i) for i in sizes):
```
Comment (pitrou) [inline dask/array/core.py]: Perhaps straighten the left part of this into `i not in (-1, 0, 1, dim)`?

**Comment by pitrou** [comment]
Would it be useful to also add a higher-level test as showcased in #2780?

File: dask/array/core.py (hunk 1 above)
Comment (jakirkham) [reply dask/array/core.py]: Yeah was thinking about that as well.

**Comment by jakirkham** [comment]
That would be a good idea. Any thoughts on where such a test should live?

**Comment by pitrou** [comment]
Perhaps somewhere in `test_array_core.py`? There's the `test_operators()` test function, otherwise you could make a new test.

**Comment by jakirkham** [comment]
Think I have addressed the concerns raised. Should be ready for another look.

**Comment by pitrou** [comment]
This looks fine to me. Thank you @jakirkham !

**Comment by jakirkham** [comment]
Thanks @pitrou.

**Pull request merged**